### DISCUSSION TITLE
Major performance issues when moving nodes after updating to v10

### DISCUSSION THREAD

**Discussion by TobiasKrok** [xyflow/xyflow#2353] [Q&A] [answered]
Hi!

I recnetly updated ny app using react-flow-renderer v9 to the latest v10.

I have now migrated everything, but I see a major decrease in performance when moving nodes using `applyNodeChanges` to update my Redux store containing my nodes/edges.
With around 50 nodes, dragging is almost like a slide show due to `applyNodeChanges` firing constantly and replacing my state (which is how it is done in the examples)

My onNodeChange looks like this:
```javascript
    const onNodeChange = useCallback((changes: NodeChange[]) => {
        dispatch(diagramCreatorSlice.actions.updateNodes(changes));

    }, []);

```

and my Redux slice method looks like this:
```javascript
updateNodes: (state: DiagramState, action: PayloadAction<NodeChange[]>) => {
            state.nodes = applyNodeChanges(action.payload, state.nodes);
        },
```

Is this the wrong way to do it? Everything has been memoized as well. It was working perfectly in v9.

**Comment by TobiasKrok** [comment, marked as answer]
Is Zustand more efficient in this manner? Such as constant updating of its store? I'm having a hard time figuring out how to solve this :(

**Reply by moklick** [reply]
No, it should be similar. Are you using the `nodes` in another component than the component where you are using `<ReactFlow />`? Are you using custom nodes that are not memoized? Most of the times the performance issues are solvable by decreasing re-renderings. If nothing helps, you can create an [uncontrolled flow](https://reactflow.dev/docs/guides/uncontrolled-flow/) by using `defaultNodes` and `defaultEdges`. In that case React Flow handles all internal state updates for you, but you can only update the nodes and edges by using the functions of the [React Flow instance](https://reactflow.dev/docs/api/react-flow-instance/).   

If you have a repo or codesandbox I am happy to take a look!

**Reply by TobiasKrok** [reply]
Yes, every custom node has been memoized and they're not used outside of the Flow. I unfortunately cannot share the repo but it looks like opting for an uncontrolled flow is the solution, however, keeping two states updated isn't ideal.  For our application at least, using onNodesChange to update our store is killing performance because each small move causes a large number of events to trigger and update the store. It also causes just as many rerenders due to redux selector 

```javascript
    const nodes = useTypedSelector((state) => state.diagram.present.nodes);
    const edges = useTypedSelector((state) => state.diagram.present.edges);
```
Ideally, it just gives me the end position (as you can get with `onNodeDragStop`) and does not update the entire store when just dragging.

**Reply by moklick** [reply]
Got it. In v9, we always had the problem with un-synced states. A partially synced state as you describe comes with different issues when you want to handle collisions for example. That's why we decided to offer a controlled and uncontrolled API. Not sure what causes the performance issues in your case.. How many nodes are you rendering?

**Reply by TobiasKrok** [reply]
Yes, the uncontrolled API fixed our problem thankfully! Our biggest diagram is at around 70 nodes. The nodes contain Ant Design components but they're all memoized.

**Reply by moklick** [reply]
With an uncontrolled flow you should be able to get better performance than with v9. However we will work on this and try to find ways how we can improve this with an uncontrolled flow too.
